### ISSUE TITLE
Multi-thread release rebuild fails

### ISSUE THREAD

**Issue by rasup** [ThrowTheSwitch/Ceedling#191]
*TL;DR*: With `:compile_threads: N` (where `N>1`) I expected `ceedling release` to build without errors but it doesn't always do that.

**How to reproduce**
```
$ ceedling new project
$ cd project
$ patch -p1 < compile_threads_issue_patch.txt
$ while ceedling release; do echo "******"; done
```
The patch, [compile_threads_issue_patch.txt](https://github.com/ThrowTheSwitch/Ceedling/files/1155856/compile_threads_issue_patch.txt), sets `:compile_threads: 2` and adds some source files.

`ceedling release` will eventually fail with the following exception:

```
Release build 'MyApp.out'
-------------------------
ERROR: Found no file 'modulea.c' in search paths.
rake aborted!

C:/compile_threads_issue/vendor/ceedling/lib/ceedling/file_finder_helper.rb:43:in `blow_up'
C:/compile_threads_issue/vendor/ceedling/lib/ceedling/file_finder_helper.rb:29:in `find_file_in_collection'
C:/compile_threads_issue/vendor/ceedling/lib/ceedling/file_finder.rb:110:in `find_compilation_input_file'
C:/compile_threads_issue/vendor/ceedling/lib/ceedling/rules_release.rake:24:in `block in <top (required)>'
C:/compile_threads_issue/vendor/ceedling/lib/ceedling/rake_wrapper.rb:18:in `[]'
C:/compile_threads_issue/vendor/ceedling/lib/ceedling/task_invoker.rb:85:in `block in invoke_release_objects'
C:/compile_threads_issue/vendor/ceedling/lib/ceedling/par_map.rb:10:in `block (2 levels) in par_map'
Tasks: TOP => release
(See full trace by running task with --trace)
```

**Other information**
This seem to only occur on rebuild i.e. never when compiling after `ceedling clean` or `ceedling clobber`.
With few source files rebuild doesn't always fail but with many files it happens all the time.
`clean` and `clobber` also fails with many files but I can't reproduce that using the attached patch.

**Environment**
I'm running the above commands in git-bash which comes with [git for windows](https://github.com/git-for-windows/git/releases).
```
$ ceedling version
Welcome to Ceedling!
  Ceedling::   0.28.1
  CMock::      2.4.4
  Unity::      2.4.1
  CException:: 1.3.1
$ ruby version
ruby 2.4.1p111 (2017-03-22 revision 58053) [x64-mingw32]
$ cmd //c ver
Microsoft Windows [Version 10.0.14393]
```

**Comment by dpostorivo** [comment]
Can you try #205 and see if it helps? I ran the same test overnight and could not get it to crash with that patch. 


Also on a related note, how much does the compiler thread option speed up the release build for you? I tried to play with it last night and implement it the test paths(before this fix), but got a negligible result on a project that takes ~ 5 minutes to test from clean & clobber.

**Comment by rasup** [comment]
> Can you try #205 and see if it helps?

It did, awesome! Thanks.

> how much does the compiler thread option speed up the release build for you?

I ran the release build a few times and got these results:

Threads | Time (approx.)
 -------- | -------
1            | 4m15s
2            | 2m25s
4            | 1m45s

I used `rake` though, and not `ceedling`, but I guess that shouldn't matter much.

I'm not sure I understand what you measured - did you run `ceedling test:...`?
Anyway, to me it looks like `:compile_threads` only affects release builds (the option only seems to be used in [task_invoker.rb](https://github.com/ThrowTheSwitch/Ceedling/blob/master/lib/ceedling/task_invoker.rb)). There is an unused option called `:test_threads` in [defaults.rb](https://github.com/ThrowTheSwitch/Ceedling/blob/master/lib/ceedling/defaults.rb#L242) should that be used for tests? And `:compile_threads` for release builds only?

No matter what, tests can be built multi threaded too by applying this patch:
```diff
diff --git a/lib/ceedling/task_invoker.rb b/lib/ceedling/task_invoker.rb
index 6bd9e6ef..adb4a0c7 100644
--- a/lib/ceedling/task_invoker.rb
+++ b/lib/ceedling/task_invoker.rb
@@ -61,12 +61,16 @@ class TaskInvoker

   def invoke_test_preprocessed_files(files)
     @dependinator.enhance_preprocesed_file_dependencies( files )
-    files.each { |file| @rake_wrapper[file].invoke }
+    par_map(PROJECT_COMPILE_THREADS, files) do |file|
+       @rake_wrapper[file].invoke
+    end
   end

   def invoke_test_dependencies_files(files)
     @dependinator.enhance_dependencies_dependencies( files )
-    files.each { |file| @rake_wrapper[file].invoke }
+    par_map(PROJECT_COMPILE_THREADS, files) do |file|
+       @rake_wrapper[file].invoke
+    end
   end

   def invoke_test_results(result)
```

I don't know how to make the test compilation part of the build multi threaded using `:compile_threads`, but as a workaround rake's multitask option can be used (`rake -m`).

**Comment by dpostorivo** [comment]
I've tried this with unfruitful results. Even with increasing the :compile_threads option and the patch above My tests still hover at ~120s from a clean build. I'm going to do a deeper investigation and see if my OS is making this feature funky.

I did see a noticeable improvement in execution time with`rake -m` going from 120s to about 104s.

**Comment by rasup** [comment]
@dpostorivo do you have `:use_deep_dependencies` set to `false` and what about `:use_test_preprocessor`? My patch only makes a difference if either of them is set to `true`.
